Parser and ArgumentTokenizer cleanups #235

PierceAndy opened the pull request:
Fixes #235 
Collecting a list of small cleanups that can be made.
Cleanups have also been applied to 97c34f2b96ec34b3afd93876f3eda12455759c91 that addresses #141.

## `ArgumentTokenizer`

- `getPreamble()`: Rather than returning `Optional.empty()`, an empty string can be returned instead. (With defensive programming, callers will need to check for empty string anyway).

- `getAllValues()`: Rather than returning `Optional.empty()`, an empty list can be returned instead. (With defensive programming, callers will need to check for an empty list anyway). This has the side benefit of removing the `toSet()` helper in `Parser`.

## `Parser`

Use of regular expressions that are less readable than their less fancy counterparts:

- `PERSON_INDEX_ARGS_FORMAT` can be replaced with `String.isEmpty()`. Actually, I think `StringUtils.isUnsignedInteger()` already checks for empty string soo...

- `KEYWORDS_ARGS_FORMAT` can be replaced with `keywords.length <= 0`.

Zhiyuan-Amos commented on src/main/java/seedu/address/logic/parser/Parser.java:
```diff
@@ -288,14 +279,13 @@ private Command prepareSelect(String args) {
     private Command prepareFind(String args) {
-        final Matcher matcher = KEYWORDS_ARGS_FORMAT.matcher(args.trim());
-        if (!matcher.matches()) {
+        if (args.trim().isEmpty()) {
```
since `args.trim()` is used twice, we can assign it to a local variable.

PierceAndy replied:
Ah, that's true. Thank you for your review.

Zhiyuan-Amos commented:
@PierceAndy Your PR is failing some tests (as reported in Continuous Integration). >< 

`guitests.EditCommandTest > edit_invalidValues_failure FAILED
    org.junit.ComparisonFailure at EditCommandTest.java:103
seedu.address.logic.LogicManagerTest > execute_add_invalidArgsFormat FAILED
    org.junit.ComparisonFailure at LogicManagerTest.java:128
129 tests completed, 2 failed`

PierceAndy commented:
Yeah, I'm working on that right now.

damithc commented:
@PierceAndy Don't forget to use CanIHasReview. More details [here](https://github.com/oss-generic/process/blob/master/docs/HowToGuides.md#create-a-multi-step-pr)

PierceAndy commented:
@damithc I was facing some authentication issues with CanIHasReview, but I've reached out to Paul and it has been resolved. It will be included as soon as possible.

pyokagan commented:
@PierceAndy Please call for at least one peer reviewer as well.

pyokagan commented on src/main/java/seedu/address/logic/parser/ArgumentTokenizer.java:
```diff
@@ -41,37 +41,31 @@ public void tokenize(String argsString) {
-        return getAllValues(prefix).flatMap((values) -> Optional.of(values.get(values.size() - 1)));
+    public String getValue(Prefix prefix) {
+        List<String> values = getAllValues(prefix);
+        return values.isEmpty() ? "" : values.get(values.size() - 1);
```
Isn't it possible to have an empty string as a value? In that case, we can't tell when looking at an empty string whether it means "no value specified" or "value of empty string specified", which is not good.

CanIHasReview commented:
# v1

@PierceAndy submitted v1 for review.

* [v1 1/3] [ArgumentTokenizer: cleanup unnecessary Optionals](https://github.com/se-edu/addressbook-level4/pull/261/commits/c9a34ba7d1b27a2918e714312376c3e9906760b8)
* [v1 2/3] [Parser: cleanup unnecessary toSet\(\) helper](https://github.com/se-edu/addressbook-level4/pull/261/commits/c9aadec8cb078c106b4665ab663f702697e34123)
* [v1 3/3] [Parser: cleanup unnecessary regex](https://github.com/se-edu/addressbook-level4/pull/261/commits/641eca177b34a64a12f231191695cafb4060bfa4)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/8164678f10e07049144c98acd52fc3572ad5c9a8...641eca177b34a64a12f231191695cafb4060bfa4))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/261/1/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

MightyCupcakes commented:
2 test failures from Travis

> guitests.EditCommandTest > edit_invalidValues_failure FAILED
> seedu.address.logic.LogicManagerTest > execute_add_invalidArgsFormat FAILED

damithc commented:
@PierceAndy need to wait for CI to pass before posting the new iteration.

PierceAndy commented:
@damithc Apologies. I accidentally dropped a commit during rebase, so CanIHasReview couldn't work. I updated the branch with the missing commit, but didn't change anything else so that CanIHasReview's v1 would be representative of the initial state.

I'm currently working on getting the CI to pass.

CanIHasReview commented:
# v2

@PierceAndy submitted v2 for review.

* [v2 1/3] [ArgumentTokenizer: cleanup unnecessary Optionals](https://github.com/se-edu/addressbook-level4/pull/261/commits/986a5357152050276a93d1c49a5d35ea72b9345a)
* [v2 2/3] [Parser: cleanup unnecessary toSet\(\) helper](https://github.com/se-edu/addressbook-level4/pull/261/commits/47413873e9edf489c121dfd72ef6cfeca38469b2)
* [v2 3/3] [Parser: cleanup unnecessary regex](https://github.com/se-edu/addressbook-level4/pull/261/commits/3ea23f6d676c3f2abcad079297e1eb5f3b13ca03)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/da4dc8e78218b35ccf991108acf1c56ca8492f57...3ea23f6d676c3f2abcad079297e1eb5f3b13ca03)) ([:chart_with_upwards_trend: Interdiff between v1 and v2](https://github.com/se-edu/addressbook-level4/commit/1bdf5c096c834ca110ce01f4960497383608be9e))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/261/2/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

PierceAndy commented:
- Toned down the excessive pruning of `Optional<String>`
- Introduced `trimmedArgs` to replace repeated instances of `args.trim()`
- Updated commit messages

MightyCupcakes commented on src/main/java/seedu/address/logic/parser/ArgumentTokenizer.java:
```diff
@@ -44,33 +45,37 @@ public void tokenize(String argsString) {
     /**
      * Returns the preamble (text before the first valid prefix), if any. Leading/trailing spaces will be trimmed.
-     *     If the string before the first prefix is empty, Optional.empty() will be returned.
+     *     If the string before the first prefix is empty, an empty string will be returned.
```
Don't think the 4 spaces in this line is necessary

PierceAndy replied:
I'm not sure why the previous contributor had that commenting style, but I think it's rampant enough to warrant its own issue+PR for cleaning up.

MightyCupcakes replied:
I am not sure, but since you modified the behaviour of the method in this PR, it is does not seem odd to modify its header comment in the same PR.

PierceAndy commented:
@yamgent @MightyCupcakes Noted. Thanks for the feedback. I'll work on improving it.

CanIHasReview commented:
# v3

@PierceAndy submitted v3 for review.

* [v3 1/4] [ArgumentTokenizer: refactor getPreamble\(\)](https://github.com/se-edu/addressbook-level4/pull/261/commits/e7def4a56da146845d8113d59d4b136255d51fed)
* [v3 2/4] [ArgumentTokenizer: refactor getAllValues\(\)](https://github.com/se-edu/addressbook-level4/pull/261/commits/2e237bbddbea6c7512ce9b046d4bb248a6f225cb)
* [v3 3/4] [Parser: refactor unnecessary toSet\(\) helper](https://github.com/se-edu/addressbook-level4/pull/261/commits/6769eaa5df4c54fa8cd4dc5cac7f4cb9dc33869b)
* [v3 4/4] [Parser: cleanup unnecessary regex](https://github.com/se-edu/addressbook-level4/pull/261/commits/a246e7ff70484d8730d721403ca56badc78e1542)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/941d0afebe52dc6bd86b60bd4a11861f48273900...a246e7ff70484d8730d721403ca56badc78e1542)) ([:chart_with_upwards_trend: Interdiff between v2 and v3](https://github.com/se-edu/addressbook-level4/commit/f17e29d2c1cfdbb94e8c6c1b69b18900fe5dce77))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/261/3/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

PierceAndy commented:
- Split first commit into two commits to be more atomic
- Improved refactoring on commit [v2 2/3] -> [v3 3/4], remove redundant logic and utilize existing methods and classes
- Improved commit messages

CanIHasReview commented:
# v4

@PierceAndy submitted v4 for review.

* [v4 1/4] [Simplify getPreamble\(\) return type](https://github.com/se-edu/addressbook-level4/pull/261/commits/1f2c16547a2c1b57d90560640402c7f705e65c89)
* [v4 2/4] [Simplify getAllValues\(\) return type](https://github.com/se-edu/addressbook-level4/pull/261/commits/a1953ab4d8b028b155f3488d22bb1163649323a0)
* [v4 3/4] [Parser: refactor unnecessary toSet\(\) helper](https://github.com/se-edu/addressbook-level4/pull/261/commits/ed581dd38f40c8bb0ec12a75b6dcc3c528b6618d)
* [v4 4/4] [Parser: cleanup unnecessary regex](https://github.com/se-edu/addressbook-level4/pull/261/commits/3af88c76ae64aca5e95db82c75e58f2e8781a361)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/941d0afebe52dc6bd86b60bd4a11861f48273900...3af88c76ae64aca5e95db82c75e58f2e8781a361)) ([:chart_with_upwards_trend: Interdiff between v3 and v4](https://github.com/se-edu/addressbook-level4/commit/f14b075d2f0bd19a07e2e1c71d58da5b61679244))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/261/4/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

PierceAndy commented:
# v4
- Updated commit messages

CanIHasReview commented:
# v5

@PierceAndy submitted v5 for review.

* [v5 1/4] [Simplify getPreamble\(\) return type](https://github.com/se-edu/addressbook-level4/pull/261/commits/607c81c63939494574818f2ee9c7c768b7939b10)
* [v5 2/4] [Simplify getAllValues\(\) return type](https://github.com/se-edu/addressbook-level4/pull/261/commits/c35192e34d3282c359df03cb2058d6f6d9cd04c4)
* [v5 3/4] [Parser: refactor unnecessary toSet\(\) helper](https://github.com/se-edu/addressbook-level4/pull/261/commits/569fa8c3e1e22aba6df5af8816aa9d5a0a42ab2b)
* [v5 4/4] [Parser: cleanup unnecessary regex](https://github.com/se-edu/addressbook-level4/pull/261/commits/2aae76d9dd7f67870fdaec5206e7598ab5cfe098)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/2221d24f2e49a9ec8a9c0d45b9dcb03694d25b0f...2aae76d9dd7f67870fdaec5206e7598ab5cfe098)) ([:chart_with_upwards_trend: Interdiff between v4 and v5](https://github.com/se-edu/addressbook-level4/compare/a6fbb5857ea4f5cce6e33a705a382e36b7806f94~1...a6fbb5857ea4f5cce6e33a705a382e36b7806f94))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/261/5/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

PierceAndy commented:
# v5
- Updated commit messages based on reviews.

@yamgent @MightyCupcakes Thanks for the feedback.

damithc commented on src/main/java/seedu/address/logic/parser/ArgumentTokenizer.java:
```diff
@@ -44,18 +45,23 @@ public void tokenize(String argsString) {
     /**
-     * Returns all values of given prefix.
+     * Returns all values of given prefix, if any.
+     * If the prefix does not exist or has no values, an empty list will be returned.
```
`an empty list will be returned` -> `returns an empty list` (to use the same tense as the previous sentence)?

damithc commented:
Calling for @pyokagan in this iteration because only minor nits remain.

damithc commented:
@PierceAndy you can go ahead with the next iteration.

CanIHasReview commented:
# v6

@PierceAndy submitted v6 for review.

* [v6 1/4] [Simplify getPreamble\(\) return type](https://github.com/se-edu/addressbook-level4/pull/261/commits/5615c6c7badeee35b25048b1636f2d1ebfde5417)
* [v6 2/4] [Simplify getAllValues\(\) return type](https://github.com/se-edu/addressbook-level4/pull/261/commits/4d073b7fec9c955237871ac9e072435b8e064623)
* [v6 3/4] [Parser: refactor unnecessary toSet\(\) helper](https://github.com/se-edu/addressbook-level4/pull/261/commits/082b8752dd525a9bed7b431696de79e56fbffbcc)
* [v6 4/4] [Parser: cleanup unnecessary regex](https://github.com/se-edu/addressbook-level4/pull/261/commits/2d150f1d555585ced4fb5bacebf9a9059eee0be5)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/13e914d0558b8b2e402910002497651ae492b046...2d150f1d555585ced4fb5bacebf9a9059eee0be5)) ([:chart_with_upwards_trend: Interdiff between v5 and v6](https://github.com/se-edu/addressbook-level4/compare/19f60e10f9332e1612aad5457027d33f6bec9d96~1...19f60e10f9332e1612aad5457027d33f6bec9d96))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/261/6/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

CanIHasReview commented:
# v7

@PierceAndy submitted v7 for review.

* [v7 1/4] [Simplify getPreamble\(\) return type](https://github.com/se-edu/addressbook-level4/pull/261/commits/0fa1b695ba52323a9b1e278696812bf99cadb342)
* [v7 2/4] [Simplify getAllValues\(\) return type](https://github.com/se-edu/addressbook-level4/pull/261/commits/d373893f651a9db09b1c3323af025d4b9bc51ccb)
* [v7 3/4] [Parser: refactor unnecessary toSet\(\) helper](https://github.com/se-edu/addressbook-level4/pull/261/commits/ccf485150c0c6d1ab1aa230bd685f2399372f234)
* [v7 4/4] [Parser: cleanup unnecessary regex](https://github.com/se-edu/addressbook-level4/pull/261/commits/da91c75e113af7d84a0d60e96e5517d59cba8755)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/13e914d0558b8b2e402910002497651ae492b046...da91c75e113af7d84a0d60e96e5517d59cba8755)) ([:chart_with_upwards_trend: Interdiff between v6 and v7](https://github.com/se-edu/addressbook-level4/compare/905b5896f97ffe0882c8c4c3a83a863b5229cee3~1...905b5896f97ffe0882c8c4c3a83a863b5229cee3))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/261/7/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

PierceAndy commented:
# v6
- Updated comments in ArgumentTokenizer

# v7
- Updated commit messages based on reviews.

@damithc Thanks for the feedback

damithc commented on src/main/java/seedu/address/logic/parser/ArgumentTokenizer.java:
```diff
@@ -44,33 +45,37 @@ public void tokenize(String argsString) {
+            return Optional.empty();
+        } else {
+            return Optional.of(values.get(values.size() - 1));
+        }
```
Use `return ... ? ... : ... ;` ?

CanIHasReview commented:
# v8

@PierceAndy submitted v8 for review.

* [v8 1/4] [Simplify getPreamble\(\) return type](https://github.com/se-edu/addressbook-level4/pull/261/commits/15ac11eb7a7461b6488475d7911570c136cc85a6)
* [v8 2/4] [Simplify getAllValues\(\) return type](https://github.com/se-edu/addressbook-level4/pull/261/commits/0cb26281db88abcb84703e2ce4309200ac77e949)
* [v8 3/4] [Parser: refactor unnecessary toSet\(\) helper](https://github.com/se-edu/addressbook-level4/pull/261/commits/827ba0b26f21baa1fc8af0472f3511e4f8cad9e0)
* [v8 4/4] [Parser: cleanup unnecessary regex](https://github.com/se-edu/addressbook-level4/pull/261/commits/d026793953ff33dfc3f86d34d78bd8bf72ff0730)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/13e914d0558b8b2e402910002497651ae492b046...d026793953ff33dfc3f86d34d78bd8bf72ff0730)) ([:chart_with_upwards_trend: Interdiff between v7 and v8](https://github.com/se-edu/addressbook-level4/compare/3cbf103046e28d43c922c6dc3395043d858ec2f3~1...3cbf103046e28d43c922c6dc3395043d858ec2f3))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/261/8/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

PierceAndy commented:
# v8
- Updated every commit message
- Used ternary operator in ArgumentTokenizer::getValue

CanIHasReview commented:
# v9

@PierceAndy submitted v9 for review.

* [v9 1/4] [Simplify getPreamble\(\) return type](https://github.com/se-edu/addressbook-level4/pull/261/commits/d003f7e3bcbef266681596335291b0eb5f8cbd23)
* [v9 2/4] [Simplify getAllValues\(\) return type](https://github.com/se-edu/addressbook-level4/pull/261/commits/cc87c04364bcdd6b9164f3275eaf1c71d2d4ef17)
* [v9 3/4] [Parser: refactor unnecessary toSet\(\) helper](https://github.com/se-edu/addressbook-level4/pull/261/commits/ab2e40094e0e7b708f783fb9c00a8444f029a4a9)
* [v9 4/4] [Parser: cleanup unnecessary regex](https://github.com/se-edu/addressbook-level4/pull/261/commits/0a47e6948b7dab9138865ae15f0ea975d2e26a22)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/13e914d0558b8b2e402910002497651ae492b046...0a47e6948b7dab9138865ae15f0ea975d2e26a22)) ([:chart_with_upwards_trend: Interdiff between v8 and v9](https://github.com/se-edu/addressbook-level4/compare/2f9f574673c1dbd80a2419daba45f5673b281592~1...2f9f574673c1dbd80a2419daba45f5673b281592))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/261/9/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

PierceAndy commented:
# v9
- Update [1/4] and [2/4] commit messages
- Use ternary operator in ArgumentTokenizer::getPreamble

pyokagan commented on src/main/java/seedu/address/logic/parser/ArgumentTokenizer.java:
```diff
@@ -59,19 +59,14 @@ public void tokenize(String argsString) {
-        } else {
-            return Optional.empty();
-        }
+        return (storedPreamble.isPresent() && !storedPreamble.get().isEmpty()) ? storedPreamble.get() : "";
```
Hrm, couldn't this just be
```java
return storedPreamble.orElse("");
```

If the `storedPreamble` just happens to contain an empty string (the `!storedPreamble.get().isEmpty()` check), the return value is still an empty string anyway.

pyokagan commented on src/main/java/seedu/address/logic/parser/Parser.java:
```diff
@@ -288,7 +288,6 @@ private Command prepareSelect(String args) {
             return Optional.empty();
         }
         return Optional.of(Integer.parseInt(index));
-
```
Unnecessary change?

pyokagan commented on src/main/java/seedu/address/logic/parser/ArgumentTokenizer.java:
```diff
@@ -44,18 +45,19 @@ public void tokenize(String argsString) {
     public Optional<String> getValue(Prefix prefix) {
-        return getAllValues(prefix).flatMap((values) -> Optional.of(values.get(values.size() - 1)));
+        List<String> values = getAllValues(prefix);
+        return values.isEmpty() ? Optional.empty() : Optional.of(values.get(values.size() - 1));
```
What's the reason for this change? Isn't this the same as the above, except less succinct?

If anything, I think it could be further shortened to:
```java
return getAllValues(prefix).map(values -> values.get(values.size() - 1));
```

pyokagan replied:
> What's the reason for this change? Isn't this the same as the above, except less succinct?

This is why I shouldn't be doing reviews at 1am in the morning.

I'll stop here.

PierceAndy commented:
@pyokagan 

> I usually like to put the main subject of the commit in front e.g. ArgumentTokenizer: simplify getPreamble() return type. But this is personal preference (afaik there's no project policy about this yet.)

[As mentioned in v4](https://github.com/se-edu/addressbook-level4/pull/261#issuecomment-276868923), `ArgumentTokenizer: simplify getPreamble() return type` [exceeds the 50 character limit for the subject line](https://github.com/oss-generic/process/blob/master/docs/FormatsAndConventions.md#commit-message).

> Need a full stop here.

Noted, thanks.

> Maybe put brackets (e.g. ArgumentTokenizer::getPreamble()) to make it clear that it is a method?

I think `Let's change the ArgumentTokenizer::getPreamble method` is clear enough. Unless the argument is against the use of Java 8 specific syntax in commit messages.

> [v9 1/4] code reviews

Noted, thanks. I don't know how I missed those.

CanIHasReview commented:
# v10

@PierceAndy submitted v10 for review.

* [v10 1/4] [Simplify getPreamble\(\) return type](https://github.com/se-edu/addressbook-level4/pull/261/commits/a0495bad65a28c675704e739c870a1c4726d53d7)
* [v10 2/4] [Simplify getAllValues\(\) return type](https://github.com/se-edu/addressbook-level4/pull/261/commits/13e470eb366af39927f48a31d95e3bd6b1e5ea80)
* [v10 3/4] [Parser: refactor unnecessary toSet\(\) helper](https://github.com/se-edu/addressbook-level4/pull/261/commits/2c2497dc7d9c1ea194db3e5a15908f725a2096e8)
* [v10 4/4] [Parser: cleanup unnecessary regex](https://github.com/se-edu/addressbook-level4/pull/261/commits/156afd23a39e2f58be65fb256fb00d13e06f1922)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/13e914d0558b8b2e402910002497651ae492b046...156afd23a39e2f58be65fb256fb00d13e06f1922)) ([:chart_with_upwards_trend: Interdiff between v9 and v10](https://github.com/se-edu/addressbook-level4/compare/5f3870eea6e7068437593da1ceb963dc232e0591~1...5f3870eea6e7068437593da1ceb963dc232e0591))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/261/10/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

PierceAndy commented:
# v10
- Update [v9 1/4] commit message
- Address [v9 1/4] code review problems
- `Let's change the ArgumentTokenizer::getPreamble method` is unchanged as per my reasoning above.

CanIHasReview commented:
# v11

@PierceAndy submitted v11 for review.

* [v11 1/4] [ArgumentTokenizer: simplify getPreamble\(\) return type](https://github.com/se-edu/addressbook-level4/pull/261/commits/098d069915c1d42500e9cb7d4b2af264da6b521a)
* [v11 2/4] [ArgumentTokenizer: simplify getAllValues\(\) return type](https://github.com/se-edu/addressbook-level4/pull/261/commits/ec55987304bf23690e8d59f380ccfc66554beb05)
* [v11 3/4] [Parser: refactor unnecessary toSet\(\) helper](https://github.com/se-edu/addressbook-level4/pull/261/commits/a078820220ee07d1fedaa711ac397a09bbf577cd)
* [v11 4/4] [Parser: cleanup unnecessary regex](https://github.com/se-edu/addressbook-level4/pull/261/commits/3dd6ebe9f5c794de2dc88f2ec1fd4f3ded2ed107)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/13e914d0558b8b2e402910002497651ae492b046...3dd6ebe9f5c794de2dc88f2ec1fd4f3ded2ed107)) ([:chart_with_upwards_trend: Interdiff between v10 and v11](https://github.com/se-edu/addressbook-level4/compare/4dae7122ebef02716cf6829e226ab20237c3d572~1...4dae7122ebef02716cf6829e226ab20237c3d572))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/261/11/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

PierceAndy commented:
# v11
- Update [1/4] and [2/4] commit messages to include `ArgumentTokenizer` at the start of the subject line. ([See: issue in oss-generic](https://github.com/oss-generic/process/issues/52))
- Update all commit messages to use `Class#method()` to refer to class methods. ([See: issue in oss-generic](https://github.com/oss-generic/process/issues/51))

CanIHasReview commented:
# v12

@PierceAndy submitted v12 for review.

* [v12 1/5] [ArgumentTokenizer: simplify getPreamble\(\) return type](https://github.com/se-edu/addressbook-level4/pull/261/commits/ca4521f8a0dc86fca9192c3ebd646bd0e420d484)
* [v12 2/5] [ArgumentTokenizer: simplify getAllValues\(\) return type](https://github.com/se-edu/addressbook-level4/pull/261/commits/f5f05c6a1d9d4bec25581ddfe591c9e70edd6c1a)
* [v12 3/5] [ParserUtil: refactor unnecessary toSet\(\)](https://github.com/se-edu/addressbook-level4/pull/261/commits/7bcc64c2046466fc3752c5da97ef7b81b0a1b8fc)
* [v12 4/5] [ParserUtil: simplify regex INDEX\_ARGS\_FORMAT](https://github.com/se-edu/addressbook-level4/pull/261/commits/dc176d886ef90d1d648af9549616390e972db55d)
* [v12 5/5] [CliSyntax: simplify regex KEYWORDS\_ARGS\_FORMAT](https://github.com/se-edu/addressbook-level4/pull/261/commits/9f9533332624096d8f2e88b701461d041074ceeb)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/aaff772f1989c11e294c264246da16453876bad1...9f9533332624096d8f2e88b701461d041074ceeb)) ([:chart_with_upwards_trend: Interdiff between v11 and v12](https://github.com/se-edu/addressbook-level4/compare/83010d5f3fe5f15eba9dedb7b9123911571d1804~1...83010d5f3fe5f15eba9dedb7b9123911571d1804))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/261/12/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

PierceAndy commented:
# v12
- Updated commit messages to comply with new proposed format.
- Split [v11 4/4] into two commits to be more atomic.
- Resolved merge conflicts with master.

MightyCupcakes commented on src/main/java/seedu/address/logic/parser/ParserUtil.java:
```diff
@@ -50,9 +49,8 @@
-    public static Set<String> toSet(Optional<List<String>> list) {
-        List<String> elements = list.orElse(Collections.emptyList());
-        return new HashSet<>(elements);
+    public static Set<String> toSet(List<String> list) {
```
What is the reason for this change?

PierceAndy replied:
It's a helper function primarily used to unwrap the Optional return value of ArgumentTokenizer#getAllValues(), which is removed in this commit.
Not changing this prevents the program from compiling.

MightyCupcakes replied:
Ah sorry. You can dismiss my review.

CanIHasReview commented:
# v13

@PierceAndy submitted v13 for review.

* [v13 1/5] [ArgumentTokenizer: simplify getPreamble\(\) return type](https://github.com/se-edu/addressbook-level4/pull/261/commits/ca4521f8a0dc86fca9192c3ebd646bd0e420d484)
* [v13 2/5] [ArgumentTokenizer: simplify getAllValues\(\) return type](https://github.com/se-edu/addressbook-level4/pull/261/commits/f5f05c6a1d9d4bec25581ddfe591c9e70edd6c1a)
* [v13 3/5] [ParserUtil: refactor unnecessary toSet\(\)](https://github.com/se-edu/addressbook-level4/pull/261/commits/7bcc64c2046466fc3752c5da97ef7b81b0a1b8fc)
* [v13 4/5] [ParserUtil: simplify regex INDEX\_ARGS\_FORMAT](https://github.com/se-edu/addressbook-level4/pull/261/commits/a7b1d30882c201a501e56f932dd8cea007af874b)
* [v13 5/5] [CliSyntax: simplify regex KEYWORDS\_ARGS\_FORMAT](https://github.com/se-edu/addressbook-level4/pull/261/commits/5f6e358c0ba4cd1034e3cbf9ca63857aad517404)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/aaff772f1989c11e294c264246da16453876bad1...5f6e358c0ba4cd1034e3cbf9ca63857aad517404)) ([:chart_with_upwards_trend: Interdiff between v12 and v13](https://github.com/se-edu/addressbook-level4/compare/b248afbf4d24302ef56bf6e6856081756a061b3c~1...b248afbf4d24302ef56bf6e6856081756a061b3c))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/261/13/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

PierceAndy commented:
# v13
- Updated [v12 4/5] - [v12 5/5] commit messages based on reviews.

pyokagan commented on src/main/java/seedu/address/logic/parser/FindCommandParser.java:
```diff
@@ -22,14 +20,14 @@
 
         // keywords delimited by whitespace
-        final String[] keywords = matcher.group("keywords").split("\\s+");
+        final String[] keywords = trimmedArgs.split("\\s+");
```
Due to how String#split() works (i.e. it trims off any empty strings at the end of the array), we might be able to just do:
```java
final String[] keywords = args.trim().split("\\s+");
if (!keywords.length) {
    return new IncorrectCommand(...);
}
...
```

damithc replied:
Yup, that would shorten the code. 👍 
I would use `keywords.length == 0` though as I've not seen the shorter form used widely in the Java world.

pyokagan replied:
@damithc 
> I would use keywords.length == 0 though as I've not seen the shorter form used widely in the Java world.

Ah, right. Actually, I don't think `!keywords.length` would even compile :-P

PierceAndy replied:
Reverting this change in v15.

The suggested code here does not work as anticipated because running `String#split()` on an empty string does not return an empty list as one might reasonably expect. Instead, a list of size 1 containing an empty string is returned. The resulting check required would be `if (keywords.length() == 1 && keywords[0].isEmpty())`.

While this might be shorter, the logic behind it less obvious. This `String#split()` behavior is not immediately obvious, and the `if` statement would warrant a comment to explain it. I do not think the trade-off is worth it, and I would argue that it decreases code readability instead.

pyokagan replied:
@PierceAndy 
> The suggested code here does not work as anticipated because running String#split() on an empty string does not return an empty list as one might reasonably expect. Instead, a list of size 1 containing an empty string is returned.

Wait, this wasn't caught by our tests?

PierceAndy replied:
It was caught by our tests when I ran them locally.

CanIHasReview commented:
# v14

@PierceAndy submitted v14 for review.

* [v14 1/5] [ArgumentTokenizer: simplify getPreamble\(\) return type](https://github.com/se-edu/addressbook-level4/pull/261/commits/17103df2d2d3dd1da4e90cff03f5295926e36516)
* [v14 2/5] [ArgumentTokenizer: simplify getAllValues\(\) return type](https://github.com/se-edu/addressbook-level4/pull/261/commits/39f02d45d8bf0821d6d21001fcb7bf4a7a5d130f)
* [v14 3/5] [ParserUtil: refactor unnecessary toSet\(\)](https://github.com/se-edu/addressbook-level4/pull/261/commits/aeb952ac303cb9abfae68bf10e1aba8bff7e6df9)
* [v14 4/5] [ParserUtil: simplify regex INDEX\_ARGS\_FORMAT](https://github.com/se-edu/addressbook-level4/pull/261/commits/a0bc46b81d947cb4b501fa6acf7c01b8010b3f05)
* [v14 5/5] [CliSyntax: simplify regex KEYWORDS\_ARGS\_FORMAT](https://github.com/se-edu/addressbook-level4/pull/261/commits/9639cc785d5f21d6c3627019504482cbfded5533)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/f40ee42cb2e29a99c9c6b8757b64b0b0f4284d67...9639cc785d5f21d6c3627019504482cbfded5533)) ([:chart_with_upwards_trend: Interdiff between v13 and v14](https://github.com/se-edu/addressbook-level4/compare/235fab7f237cba8bf7dc696c48ab26312b5d4350~1...235fab7f237cba8bf7dc696c48ab26312b5d4350))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/261/14/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

PierceAndy commented:
# v14
- Shortened code in [v13 5/5]

pyokagan commented on src/main/java/seedu/address/logic/parser/ArgumentTokenizer.java:
```diff
@@ -59,19 +59,14 @@ public void tokenize(String argsString) {
-     * Returns the preamble (text before the first valid prefix), if any. Leading/trailing spaces will be trimmed.
-     *     If the string before the first prefix is empty, Optional.empty() will be returned.
+     * Returns the preamble (text before the first valid prefix), if any. Trims any leading/trailing spaces.
+     * If the string before the first prefix is empty, returns an empty string.
```
I do slightly wonder whether this is a redundant comment. It's like saying "if we have an empty string, we return an empty string".

pyokagan commented on src/test/java/seedu/address/logic/parser/ArgumentTokenizerTest.java:
```diff
@@ -1,9 +1,9 @@
-import static org.junit.Assert.assertTrue;
 import static org.junit.Assert.assertFalse;
 import static org.junit.Assert.assertNotEquals;
+import static org.junit.Assert.assertTrue;
```
Hrrm, I'm not very sure why there is such a change, actually. Is our checkstyle configuration not checking that the ordering of `import static` is in lexicographical order?

pyokagan replied:
Ah I see, this was added in [v14 1/5] and shifted in [v14 2/5].

PierceAndy replied:
Noted. Rebasing this to [1/5] in next iteration.

CanIHasReview commented:
# v15

@PierceAndy submitted v15 for review.

* [v15 1/5] [ArgumentTokenizer: simplify getPreamble\(\) return type](https://github.com/se-edu/addressbook-level4/pull/261/commits/0bdcb14c157edf1d1ca68a3bf44d9fa5ee71454f)
* [v15 2/5] [ArgumentTokenizer: simplify getAllValues\(\) return type](https://github.com/se-edu/addressbook-level4/pull/261/commits/573795f95c95da1d4cdfed46b1994f0dde5e368c)
* [v15 3/5] [ParserUtil: remove unnecessary toSet\(\)](https://github.com/se-edu/addressbook-level4/pull/261/commits/0457ddf0dcee39792e367e5e64ad1578c591cccd)
* [v15 4/5] [ParserUtil: remove unnecessary regex usage](https://github.com/se-edu/addressbook-level4/pull/261/commits/5cb3ed90413b3f01e16e22ddfc788af33912ebe6)
* [v15 5/5] [FindCommandParser: remove unnecessary regex usage](https://github.com/se-edu/addressbook-level4/pull/261/commits/99e4f379175ca9c82219832832947878314b11b6)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/f40ee42cb2e29a99c9c6b8757b64b0b0f4284d67...99e4f379175ca9c82219832832947878314b11b6)) ([:chart_with_upwards_trend: Interdiff between v14 and v15](https://github.com/se-edu/addressbook-level4/compare/027f573ef9797f9c53bf55896baf91510d2272c9~1...027f573ef9797f9c53bf55896baf91510d2272c9))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/261/15/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

PierceAndy commented:
# v15
- Reverting [change from v13 review](https://github.com/se-edu/addressbook-level4/pull/261#pullrequestreview-21687069)
- Removed redundant sentence in [v14 1/5] comments
- Corrected order of `ArgumentTokenizerTest` import statements in between [v14 1/5] & [v14 2/5]
- Updated [v14 3/5] - [v14 5/5] commit messages

pyokagan commented:
Rest of the commits are fine.

PierceAndy commented on src/test/java/seedu/address/logic/parser/ArgumentTokenizerTest.java:
```diff
@@ -33,11 +34,11 @@ public void tokenize_emptyArgsString_noValues() {
+        assertEquals(expectedPreamble, argsTokenizer.getPreamble());
     }
 
     private void assertPreambleAbsent(ArgumentTokenizer argsTokenizer) {
```
I believe it asserting that the given expected preamble can be found in the given `argsTokenizer`. I'm not sure how this is asserting that a preamble is empty.

pyokagan replied:
@PierceAndy 
> I believe it asserting that the given expected preamble can be found in the given argsTokenizer.

Is this in reference to `assertPreamblePresent()`? I was referring to `assertPreambleAbsent()`.

PierceAndy replied:
Oops my bad, I looked at the pretty colours instead.

CanIHasReview commented:
# v16

@PierceAndy submitted v16 for review.

* [v16 1/5] [ArgumentTokenizer: simplify getPreamble\(\) return type](https://github.com/se-edu/addressbook-level4/pull/261/commits/8f90d6fd4f4abd389048682afaaf54cafb119641)
* [v16 2/5] [ArgumentTokenizer: simplify getAllValues\(\) return type](https://github.com/se-edu/addressbook-level4/pull/261/commits/33e6c960509e2d0d54533cf64f7da5f403eaa832)
* [v16 3/5] [ParserUtil: remove unnecessary toSet\(\)](https://github.com/se-edu/addressbook-level4/pull/261/commits/5293a6a122aeb30dfbb5cd88143e4e90822114e0)
* [v16 4/5] [ParserUtil: remove unnecessary regex usage](https://github.com/se-edu/addressbook-level4/pull/261/commits/5b134874cd8056807d314a33ad7e85ab2cd0876c)
* [v16 5/5] [FindCommandParser: remove unnecessary regex usage](https://github.com/se-edu/addressbook-level4/pull/261/commits/56cf4fbf75d03e7e88509888c3f1f3bf3ca4f108)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/d741336a16662b675c117ed2aa3ade11a934f24c...56cf4fbf75d03e7e88509888c3f1f3bf3ca4f108)) ([:chart_with_upwards_trend: Interdiff between v15 and v16](https://github.com/se-edu/addressbook-level4/compare/24a2670c364af4b6c77f8a16612590c7865823fb~1...24a2670c364af4b6c77f8a16612590c7865823fb))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/261/16/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

PierceAndy commented:
# v16
- In `ArgumentTokenizerTest`, `assertPreambleAbsent()` -> `assertPreambleEmpty()`
- Removed unnecessary `if any` in comment in `ArgumentTokenizer`

CanIHasReview commented:
# v17

@PierceAndy submitted v17 for review.

* [v17 1/5] [ArgumentTokenizer: simplify getPreamble\(\) return type](https://github.com/se-edu/addressbook-level4/pull/261/commits/8f90d6fd4f4abd389048682afaaf54cafb119641)
* [v17 2/5] [ArgumentTokenizer: simplify getAllValues\(\) return type](https://github.com/se-edu/addressbook-level4/pull/261/commits/33e6c960509e2d0d54533cf64f7da5f403eaa832)
* [v17 3/5] [ParserUtil: remove unnecessary toSet\(\)](https://github.com/se-edu/addressbook-level4/pull/261/commits/5293a6a122aeb30dfbb5cd88143e4e90822114e0)
* [v17 4/5] [ParserUtil: remove unnecessary regex usage](https://github.com/se-edu/addressbook-level4/pull/261/commits/38cfb278c20b403950d4ae74bcdce6a94a41c7ab)
* [v17 5/5] [FindCommandParser: remove unnecessary regex usage](https://github.com/se-edu/addressbook-level4/pull/261/commits/8f9373723267ee88fc5a1e07fbc421c0cd91f122)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/d741336a16662b675c117ed2aa3ade11a934f24c...8f9373723267ee88fc5a1e07fbc421c0cd91f122)) ([:chart_with_upwards_trend: Interdiff between v16 and v17](https://github.com/se-edu/addressbook-level4/compare/3444c51db0562a7d3439b87ef6337913618bfb9d~1...3444c51db0562a7d3439b87ef6337913618bfb9d))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/261/17/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

PierceAndy commented:
# v18
- Updated [4/5] & [5/5] commit messages

pyokagan commented on src/main/java/seedu/address/logic/parser/ArgumentTokenizer.java:
```diff
@@ -59,19 +59,13 @@ public void tokenize(String argsString) {
+    public String getPreamble() {
         Optional<String> storedPreamble = getValue(new Prefix(""));
 
         /* An empty preamble is considered 'no preamble present' */
```
This comment is actually referring to the `storedPreamble.isPresent() && !storedPreamble.get().isEmpty()` line below. Given that you are changing the logic here, this comment needs to be updated, or maybe even completely removed. (since comments that do not exist are better than outdated comments)

FWIW, I took a closer glance at the `ArgumentTokenizer` code, and it appears that `storedPreamble` will never be `Optional.empty()` if the `tokenize()` method is called. Only when the `tokenize()` method is not called will `storedPreamble` be `Optional.empty()`. Just an observation that may or may not be relevant.

pyokagan replied:
Yeah that works. (Note: I didn't actually strictly verify if the second point is true.)

PierceAndy replied:
I took a look at the code, and it seems to be right.

PierceAndy replied:
However, applying assertions here should probably have another issue of its own, since it's not strictly part of this PR. `getValue()` and other similar methods in `ArgumentTokenizer` could benefit from that as well.

pyokagan replied:
Yes, two second thoughts:

* Should probably be throwing an `IllegalStateException` rather than an assertion, since this is a public method.

* The other methods of `ArgumentTokenizer` (e.g. `getValue()`, `getAllValues()`) do not throw exceptions when called before `tokenize()` is called, so `getPreamble()` shouldn't be throwing errors to be consistent with the rest of the methods (least surprise).

PierceAndy replied:
- Yeah I just noticed that as well.
- To be consistent, all of these functions could be updated to throw an `IllegalStateException` as mentioned in #312?

CanIHasReview commented:
# v18

@PierceAndy submitted v18 for review.

* [v18 1/5] [ArgumentTokenizer: simplify getPreamble\(\) return type](https://github.com/se-edu/addressbook-level4/pull/261/commits/585b1cfc7423e63aadf15cb53ca787021ee2aa7c)
* [v18 2/5] [ArgumentTokenizer: simplify getAllValues\(\) return type](https://github.com/se-edu/addressbook-level4/pull/261/commits/46497dc8d01bcd726073ffa2895002c3c6f0ed90)
* [v18 3/5] [ParserUtil: remove unnecessary toSet\(\)](https://github.com/se-edu/addressbook-level4/pull/261/commits/21410adb16208a3802302b9dad91e94c8dda99fc)
* [v18 4/5] [ParserUtil: remove unnecessary regex usage](https://github.com/se-edu/addressbook-level4/pull/261/commits/1d46f717fd7acb2fb5ef9948733e1747713cd89e)
* [v18 5/5] [FindCommandParser: remove unnecessary regex usage](https://github.com/se-edu/addressbook-level4/pull/261/commits/da90e68f0c36fe10bc114a862d06f4eb50e08798)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/d741336a16662b675c117ed2aa3ade11a934f24c...da90e68f0c36fe10bc114a862d06f4eb50e08798)) ([:chart_with_upwards_trend: Interdiff between v17 and v18](https://github.com/se-edu/addressbook-level4/compare/f6e2248c4e70cd53205c425879dd8e2ed0f1efe7~1...f6e2248c4e70cd53205c425879dd8e2ed0f1efe7))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/261/18/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

PierceAndy commented:
# v18
- Removed outdated comment.
